src/sms/core/events_handler.py

```diff
@@ -45,17 +46,63 @@ async def get_admin_client(self) -> AsyncGenerator[AIOKafkaAdminClient, None]:
         finally:
             await admin_client.close()
 
+    async def _get_cleanup_policy(
+        self, *, admin_client: AIOKafkaAdminClient, topic: str
+    ):
+        """Get the current cleanup policy for a topic"""
+        config_response = await admin_client.describe_configs(
+            [ConfigResource(ConfigResourceType.TOPIC, name=topic)]
+        )
+        config_resources = config_response[0].resources[0]
+        configs = config_resources[-1]
+        policy = "compact"
+        for item in configs:
+            if item[0] == "cleanup.policy":
+                policy = item[1]
+                break
+        return policy
+
+    async def _set_cleanup_policy(
+        self, *, admin_client: AIOKafkaAdminClient, topic: str, policy: str
+    ):
+        """Set the cleanup policy"""
+        await admin_client.alter_configs(
+            config_resources=[
+                ConfigResource(
+                    ConfigResourceType.TOPIC,
+                    name=topic,
+                    configs={
+                        "cleanup.policy": policy,
+                    },
+                )
+            ]
+        )
+
     async def clear_topics(self, *, topics: list[str], exclude_internal: bool = True):
         """Clear messages from given topic(s).
 
         If no topics are specified, all topics will be cleared, except internal topics
         unless otherwise specified.
+
+        To enable topic clearing for compacted topics, the cleanup policy is temporarily
+        set to 'delete' while the action is performed. Afterward, the policy is set to
+        its original value (either 'compact' or 'compact,delete').
         """
         async with self.get_admin_client() as admin_client:
             if not topics:
                 topics = await admin_client.list_topics()
             if exclude_internal:
                 topics = [topic for topic in topics if not topic.startswith("__")]
+            original_policies = {}
+            for topic in topics:
+                policy = await self._get_cleanup_policy(
+                    admin_client=admin_client, topic=topic
+                )
+                original_policies[topic] = policy
```

Comment:
If you set the `original_policies` entry only in the if-block, then later you need to only run through the items of this dict when restoring everything, not through all topics.